File: 0010.RegularExpressionMatching.py
#coding: utf-8
'''
https://leetcode.com/discuss/99229/leetbook-2种思路中文详（附解释，图示，代码）
'''

class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        dp = [[False]*(len(s)+1) for _ in xrange(len(p)+1)]
        dp[0][0] = True
        for i in xrange(2, len(p)+1):
            dp[i][0] = dp[i-2][0] and p[i-1] == '*'
        for i in xrange(1, len(p)+1):
            for j in xrange(1, len(s)+1):
                if p[i-1] != '*':
                    dp[i][j] = dp[i-1][j-1] and (p[i-1] == s[j-1] or p[i-1] == '.')
                else:
                    dp[i][j] = dp[i-2][j] or dp[i-1][j]
                    if p[i-2] == s[j-1] or p[i-2] == '.':
                        dp[i][j] |= dp[i][j-1]
        return dp[-1][-1]

test = Solution()
print test.isMatch('a', 'a*')

File: 0010.RegularExpressionMatching.go
// Similar with 44. Wildcard Matching 
// https://leetcode.com/problems/regular-expression-matching/discuss/5651/Easy-DP-Java-Solution-with-detailed-Explanation
func isMatch(s string, p string) bool {
    m, n := len(s), len(p)
    dp := make([][]bool, m+1)
    for i := range dp {
        dp[i] = make([]bool, n+1)
    }
    dp[0][0] = true
    for j := 1; j <= n; j++ {
        dp[0][j] = (p[j-1] == '*' && dp[0][j-2])
    }
    for i := 1; i <= m; i++ {
        for j := 1; j <= n; j++ {
            if s[i-1] == p[j-1] || p[j-1] == '.' {
                dp[i][j] = dp[i-1][j-1]
            } else if p[j-1] == '*' {
                if s[i-1] == p[j-2] || p[j-2] == '.' {
                    dp[i][j] = dp[i-1][j] || dp[i][j-1] || dp[i][j-2]
                } else {
                    dp[i][j] = dp[i][j-2]
                }
            }
        }
    }
    return dp[m][n]
}